src/Middleware/RequestDecompression/src/DeflateDecompressionProvider.cs

```diff
@@ -13,6 +14,221 @@ internal sealed class DeflateDecompressionProvider : IDecompressionProvider
+                    return -1;
+                }
+
+                _firstByte = (byte)value;
+                _firstByteStatus = FirstByteStatus.Available;
+                return value;
+            }
+
+            public async ValueTask<int> PeekFirstByteAsync(CancellationToken cancellationToken)
+            {
+                Debug.Assert(_firstByteStatus == FirstByteStatus.None);
+
+                var buffer = new byte[1];
+
+                int bytesRead = await _stream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+                if (bytesRead == 0)
+                {
+                    _firstByteStatus = FirstByteStatus.Consumed;
+                    return -1;
+                }
+
+                _firstByte = buffer[0];
+                _firstByteStatus = FirstByteStatus.Available;
+                return buffer[0];
+            }
+
+            public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+            {
+                if (_firstByteStatus == FirstByteStatus.Available)
+                {
+                    if (buffer.Length != 0)
+                    {
+                        buffer.Span[0] = _firstByte;
+                        _firstByteStatus = FirstByteStatus.Consumed;
+                        return new ValueTask<int>(1);
+                    }
+
+                    return new ValueTask<int>(0);
+                }
+
+                Debug.Assert(_firstByteStatus == FirstByteStatus.Consumed);
+                return _stream.ReadAsync(buffer, cancellationToken);
+            }
+
+            public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                Debug.Assert(_firstByteStatus != FirstByteStatus.None);
+
+                ValidateCopyToArguments(destination, bufferSize);
+                if (_firstByteStatus == FirstByteStatus.Available)
+                {
+                    await destination.WriteAsync(new byte[] { _firstByte }, cancellationToken).ConfigureAwait(false);
+                    _firstByteStatus = FirstByteStatus.Consumed;
+                }
+
+                await _stream.CopyToAsync(destination, bufferSize, cancellationToken).ConfigureAwait(false);
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                if (_firstByteStatus == FirstByteStatus.Available)
+                {
+                    if (buffer.Length != 0)
+                    {
+                        buffer[0] = _firstByte;
+                        _firstByteStatus = FirstByteStatus.Consumed;
+                        return 1;
+                    }
+
+                    return 0;
+                }
+
+                Debug.Assert(_firstByteStatus == FirstByteStatus.Consumed);
+                return _stream.Read(buffer, offset, count);
+            }
+
+            private enum FirstByteStatus : byte
```

Comment:
Setting this to `byte` is weird. There isn't any reason to override the default since the enum field is only being read/written using the actual enum type.

Reply:
Setting it to byte makes it one byte instead of four bytes. And it's stored as a field on the stream type along with the byte that's read. With padding, on 32-bit, it makes the allocated stream 4 bytes smaller.